sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1595 @@
+burnchain height `block-height`.
+
+* The `pox-addrs` property returns a tuple with two items: a list of up to two PoX addresses that received a PoX payout at that block height, and the amount of burnchain
+tokens paid to each address (note that per the blockchain consensus rules, each PoX payout will be the same for each address in the block-commit transaction).
+The list will include burn addresses -- that is, the unspendable addresses that miners pay to when there are no PoX addresses left to be paid.  During the prepare phase,
+there will be exactly one burn address reported. During the reward phase, up to two burn addresses may be reported in the event that some PoX reward slots are not claimed.
+
+The `addrs` list contains the same PoX address values passed into the PoX smart contract:
+   * They each have type signature `(tuple (hashbytes (buff 32)) (version (buff 1)))`
+   * The `version` field can be any of the following:
+      * `0x00` means this is a pay-to-public-key-hash (p2pkh) address, and `hashbytes` is the 20-byte hash160 of a single public key
+      * `0x01` means this is a pay-to-script-hash (p2sh) address, and `hashbytes` is the 20-byte hash160 of a redeemScript script
+      * `0x02` means this is a pay-to-witness-public-key-hash-over-pay-to-script-hash (p2wpkh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wpkh witness script
+      * `0x03` means this is a pay-to-witness-script-hash-over-pay-to-script-hash (p2wsh-p2sh) address, and `hashbytes` is the 20-byte hash160 of a p2wsh witness script
+      * `0x04` means this is a pay-to-witness-public-key-hash (p2wpkh) address, and `hashbytes` is the 20-byte hash160 of the witness script
+      * `0x05` means this is a pay-to-witness-script-hash (p2wsh) address, and `hashbytes` is the 32-byte sha256 of the witness script
+      * `0x06` means this is a pay-to-taproot (p2tr) address, and `hashbytes` is the 32-byte sha256 of the witness script
+
+**Rationale:**  This method empowers developers to query Bitcoin state.  Stacks
+2.1 adds the ability for smart contracts to query PoX payouts directly, with applications
+for Stacking pools and other Stacking-centric programs.
+
+**Examples:**
+
+```clarity
+(get-burn-block-info? header-hash u677050) ;; Returns (some 0xe67141016c88a7f1203eca0b4312f2ed141531f59303a1c267d7d83ab6b977d8)
+(get-burn-block-info? pox-addrs u677050) ;; Returns (some (tuple (addrs ((tuple (hashbytes 0x395f3643cea07ec4eec73b4d9a973dcce56b9bf1) (version 0x00)) (tuple (hashbytes 0x7c6775e20e3e938d2d7e9d79ac310108ba501ddb) (version 0x01)))) (payout u123)))
+```
+
+### New method: `slice`
+
+* **Input Signature:** `(slice (sequence sequence_A) (left uint) (right uint))`
+* **Output Signature:** `(optional sequence_A)`
+
+The `slice` function attempts to return a sub-sequence of that starts
+at `left-position` (inclusive), and ends at `right-position`
```

Comment:
I will add a note, but I don't know yet if it makes sense to switch to 1-based indexing (especially since so many other popular programming languages use 0-based indexing).

Reply:
Discussing the intent of Clarity contracts with non-programmers will benefit from one-based indexing. Zero-based indexing is typically used in systems languages, while many high-level domain-specific languages use one-based indexing. Some languages using one-based indexing include Julia, Lua, Erlang, Smalltalk, Mathematica, Matlab, and R. https://github.com/clarity-lang/reference/issues/43

Reply:
As index-of and element-at are zero-based, it makes sense to use zero-based values here as well. Otherwise, we need to redefine more than what is in scope of this SIP. Maybe something for Clarity 3 :-)